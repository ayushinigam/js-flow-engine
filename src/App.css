.app {
  width: 50vw;
  min-width: 350px;
  margin: auto;
  font-family: 'Roboto', sans-serif;
  font-size: 14px;
  padding: 20px 0; }

body {
  background-color: #DBDBDB;
  min-height: 100vh; }

* {
  box-sizing: border-box; }
