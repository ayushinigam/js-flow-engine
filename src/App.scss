@import "./variables.scss";

.app {
  width: 50vw;
  min-width: 350px;
  margin: auto;
  font-family: 'Roboto', sans-serif;
  font-size: $APP_FONT_SIZE;
  padding: 20px 0;
}
body {
  background-color: $DEFAULT_BG;
  min-height: 100vh;
}

* {
  box-sizing: border-box;
}
