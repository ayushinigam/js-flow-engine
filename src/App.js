import React, { Component } from 'react';
import './App.css';
import rules from './rules.json';
import makeSafeEval from 'cross-safe-eval';
import Rule from './components/Rule/Rule.component';
import '../node_modules/reset-css/reset.css';
import { ToastContainer, toast } from 'react-toastify';

const safeEval = makeSafeEval(); // uses eval without access to globals, scope, closure and "this"

class App extends Component {
  state = {
    rules: rules,
    incomingData: '{"value": -1}',
    statuses: {}
  }
  currentCallStack = {}
  setIncomingData = (e) => this.setState({incomingData: e.target.value});

  processRule = (currentRuleID) => {
    const currentRule = this.state.rules.filter((ruleDetails) => ruleDetails.id === currentRuleID)[0];
    const ruleLogic = safeEval(`(${currentRule.body})`);
    const incomingData = JSON.parse(this.state.incomingData);
    const ruleOutput = ruleLogic(incomingData);
    const nextRuleID =  !!ruleOutput ? currentRule.passedRuleID : currentRule.failedRuleID;
    return {ruleOutput, nextRuleID};
  }

  runRules = (currentRuleID) => {
    try {
      if(this.currentCallStack.hasOwnProperty(currentRuleID)) {
        const callStack = [...Object.keys(this.currentCallStack), currentRuleID].join(' --> ');
        throw new Error(`Infinite loop at RuleID ${currentRuleID} Callstack: ${callStack}`);
      }
      const {ruleOutput, nextRuleID} = this.processRule(currentRuleID);
      this.currentCallStack[currentRuleID] = ruleOutput;
      if(nextRuleID && nextRuleID !== null) { //
        this.runRules(nextRuleID);
      }
      else { // last rule
        toast.success('✅ End of the flow!');
      }
    }
    catch(e) {
      toast.error(`⚠️ ERROR EXECUTING RULE: "${currentRuleID}". ERROR: ${e}`, {autoClose: 7500});
    }
  }

  runHandler = () => {
    this.currentCallStack = {};
    this.setState({statuses: this.currentCallStack}, () => { //clear previous state
      toast.dismiss(); //dismiss previous toasts
      this.runRules(this.state.rules[0].id);
      this.setState({statuses: this.currentCallStack})
    });

  }

  render() {
    return (
      <div className="app">
      Incoming data: <input onChange={this.setIncomingData} value={this.state.incomingData} />
      <button onClick={this.runHandler}>run rules</button>
        {this.state.rules.map((ruleDetails) => (
          <Rule key={ruleDetails.id} status={this.state.statuses[ruleDetails.id]} {...ruleDetails} />
        ))}
        <ToastContainer />
      </div>
    );
  }
}

export default App;
