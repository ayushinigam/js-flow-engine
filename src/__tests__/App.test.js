import React from 'react';
import renderer from 'react-test-renderer';
import App from '../App';
import {shallow} from 'enzyme';
import { toast } from 'react-toastify';

jest.mock('react-toastify')

describe('App component', () => {
  let wrapper, instance;
  beforeEach(()=> {
    wrapper = shallow(<App />);
    instance = wrapper.instance();
    const rules = [{
      id: 'test123',
      body: 'function(obj) {return true}',
      passedRuleID: 'passed123',
      failedRuleID: 'failed123'
    }];
    wrapper.setState({rules});
  })
  it('App: renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('setIncomingData: should set state.incomingData', () => {
    const eventData = {target: {value: '"{test: 123}"'}};
    instance.setIncomingData(eventData);
    expect(instance.state.incomingData).toBe('"{test: 123}"');
  })
  it('processRule success: should process the rule and return {ruleOutput, nextRuleID} ', () => {
    const {ruleOutput, nextRuleID} = instance.processRule('test123');
    expect(ruleOutput).toBe(true);
    expect(nextRuleID).toBe('passed123');
  })
  it('processRule failure: should process the rule and return {ruleOutput, nextRuleID} ', () => {
    const rules = [{
      id: 'xyz123',
      body: 'function(obj) {return false}',
      passedRuleID: 'passed123',
      failedRuleID: 'failed123'
    }];
    wrapper.setState({rules});
    const {ruleOutput, nextRuleID} = instance.processRule('xyz123');
    expect(ruleOutput).toBe(false);
    expect(nextRuleID).toBe('failed123');
  })

  it('runHandler: should reset and start rules with first id', () =>{
    instance.runRules = jest.fn();
    instance.runHandler();
    expect(toast.dismiss).toBeCalledWith();
    expect(instance.runRules).toHaveBeenLastCalledWith('test123');
    expect(instance.state.statuses).toEqual({})
  })
  it('should show error if rule is already executed', () => {
    instance.currentCallStack = {'abc': true};
    instance.runRules('abc');
    expect(toast.error).toHaveBeenLastCalledWith(`⚠️ ERROR EXECUTING RULE: \"abc\". ERROR: Error: Infinite loop at RuleID abc Callstack: abc --> abc`, {autoClose: 7500})
  })

  it('runRules: should recursively run rules based on nextRuleID', () => {
    const rules = [{
      id: 'xyz123',
      body: 'function(obj) {return true}',
      passedRuleID: 'passed123',
      failedRuleID: 'failed123'
    },{
      id: 'passed123',
      body: 'function(obj) {return false}',
      passedRuleID: '',
      failedRuleID: null
    }];
    wrapper.setState({rules});
    jest.spyOn(instance, 'runRules');
    instance.runRules('xyz123');
    expect(instance.runRules).toHaveBeenCalledTimes(2);
    expect(toast.success).toHaveBeenLastCalledWith('✅ End of the flow!')
  })
});
