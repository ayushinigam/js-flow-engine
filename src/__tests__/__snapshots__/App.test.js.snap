// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App: renders correctly 1`] = `
<div
  className="app"
>
  Incoming data: 
  <input
    onChange={[Function]}
    value="{\\"value\\": -1}"
  />
  <button
    onClick={[Function]}
  >
    run rules
  </button>
  <div
    className="rule-container"
  >
    <div
      className="rule-header "
      onClick={[Function]}
    >
      <span
        className="rule-id"
      >
        1
      </span>
      <p
        className="rule-title"
      >
        Should be truthy
      </p>
    </div>
    <div
      className="rule-content rule-collapsed"
    >
      <div
        className="input-with-label undefined"
      >
        <p>
          Rule body
        </p>
        <input
          className="rule-body"
          disabled={true}
          value="function (obj) {return !!obj}"
        />
      </div>
      <div
        className="next-rules-container"
      >
        <div
          className="input-with-label next-rule"
        >
          <p>
            Next rule-id if passed
          </p>
          <input
            className={undefined}
            disabled={true}
            value="2"
          />
        </div>
        <div
          className="input-with-label next-rule"
        >
          <p>
            Next rule-id if failed
          </p>
          <input
            className={undefined}
            disabled={true}
            value="4"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="rule-container"
  >
    <div
      className="rule-header "
      onClick={[Function]}
    >
      <span
        className="rule-id"
      >
        2
      </span>
      <p
        className="rule-title"
      >
        Should have a key called "value"
      </p>
    </div>
    <div
      className="rule-content rule-collapsed"
    >
      <div
        className="input-with-label undefined"
      >
        <p>
          Rule body
        </p>
        <input
          className="rule-body"
          disabled={true}
          value="function (obj) {return obj.hasOwnProperty('value')}"
        />
      </div>
      <div
        className="next-rules-container"
      >
        <div
          className="input-with-label next-rule"
        >
          <p>
            Next rule-id if passed
          </p>
          <input
            className={undefined}
            disabled={true}
            value="3"
          />
        </div>
        <div
          className="input-with-label next-rule"
        >
          <p>
            Next rule-id if failed
          </p>
          <input
            className={undefined}
            disabled={true}
            value="4"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="rule-container"
  >
    <div
      className="rule-header "
      onClick={[Function]}
    >
      <span
        className="rule-id"
      >
        3
      </span>
      <p
        className="rule-title"
      >
        obj.value should be a number
      </p>
    </div>
    <div
      className="rule-content rule-collapsed"
    >
      <div
        className="input-with-label undefined"
      >
        <p>
          Rule body
        </p>
        <input
          className="rule-body"
          disabled={true}
          value="function (obj) {return typeof obj.value === 'number'}"
        />
      </div>
      <div
        className="next-rules-container"
      >
        <div
          className="input-with-label next-rule"
        >
          <p>
            Next rule-id if passed
          </p>
          <input
            className={undefined}
            disabled={true}
            value="4"
          />
        </div>
        <div
          className="input-with-label next-rule"
        >
          <p>
            Next rule-id if failed
          </p>
          <input
            className={undefined}
            disabled={true}
            value="null"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="rule-container"
  >
    <div
      className="rule-header "
      onClick={[Function]}
    >
      <span
        className="rule-id"
      >
        4
      </span>
      <p
        className="rule-title"
      >
        obj.value should be greater than 0
      </p>
    </div>
    <div
      className="rule-content rule-collapsed"
    >
      <div
        className="input-with-label undefined"
      >
        <p>
          Rule body
        </p>
        <input
          className="rule-body"
          disabled={true}
          value="function (obj) {return obj.value > 0}"
        />
      </div>
      <div
        className="next-rules-container"
      >
        <div
          className="input-with-label next-rule"
        >
          <p>
            Next rule-id if passed
          </p>
          <input
            className={undefined}
            disabled={true}
            value="null"
          />
        </div>
        <div
          className="input-with-label next-rule"
        >
          <p>
            Next rule-id if failed
          </p>
          <input
            className={undefined}
            disabled={true}
            value="null"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
